docs/cip/CIP-07102024_Write_Ahead_Log_Pruning_Vacuuming.md

```diff
@@ -0,0 +1,65 @@
+# CIP-07102024: Write-Ahead Log Pruning & Vacuuming
+
+## Status
+
+Current Status: `Under Discussion`
+
+## Motivation
+
+Chroma's SQLite-based write-ahead log grows infinitely over time. When ingesting large amounts of data, it's not uncommon for the SQLite database to grow to many gigabytes in size. Large databases cost more, take longer to back up, and can result in decreased query performance.
+
+There are two separate problems:
+
+- The database, specifically the `embeddings_queue` table, has unbounded growth.
+- The SQLite `VACUUM` command, often recommended for such scenarios, is a blocking and potentially slow operation. Read and write operations are both blocked during a `VACUUM`.[^1]
+
+This CIP addresses both issues.
+
+## Proposed Changes
+
+A new configuration parameter will be added, `log:vacuum_threshold`. It defaults to 1GB. Following Postgres' convention, the unit is megabytes. This helps avoid excessive fragmentation—without this parameter, or if it's set to `0`, it is effectively the same as SQLite's full vacuum mode.
+
+Two additional things will be done after write transactions:
+
+1. The `embeddings_queue` table will be pruned to remove rows that are no longer needed. Specifically, rows with a sequence ID less than the minimum sequence ID of any active subscriber will be deleted.
+2. `PRAGMA freelist_count` will be checked to see if the number of free pages multiplied by the page size is greater than the `log:vacuum_threshold` parameter. If so, `PRAGMA incremental_vacuum` is run to free up pages. This is a non-blocking operation.
+
+Tentatively this will be done after every write transaction, assuming the added latency is minimal. If this is not the case, it will be run on some interval.
+
+## Public Interfaces
+
+In addition to the configuration parameter described above, a new `chroma vacuum` command will be added to the CLI to manually perform a full vacuum of the database. Usage:
+
+```bash
+chroma vacuum --path ./chroma_data
+```
+
+This automatically runs the pruning operation described above before running `VACUUM`. Prior to any modifications, it checks that there is enough available disk space to complete the vacuum (i.e. the free space on the disk is at least twice the size of the database).[^2]
+
+`chroma vacuum` should be run infrequently; it may increase query performance but the degree to which it does so is currently unknown.
```

Comment:
CLI is great, care must be taken to explain to users that if Chroma is running (e.g. writing) and this is run it has the potential to corrupt the DB

Reply:
Under the testing section I expand on this a little, I'm hoping to make that scenario impossible--maybe through file locking? But I suppose this could still happen if an older version of Chroma is running, so will add a note. 👍

Reply:
82df8ee0